engines/sci/engine/klists.cpp

```diff
@@ -824,6 +827,13 @@ reg_t kArraySetElements(EngineState *s, int argc, reg_t *argv) {
 }
 
 reg_t kArrayFree(EngineState *s, int argc, reg_t *argv) {
+	if (s->_segMan->getSegmentType(argv[0].getSegment()) == SEG_TYPE_SCRIPT) {
```

Comment:
> No, this function is not properly named. It frees a single ELEMENT of an array, not the whole array: [code]

That code doesn’t do what you are saying at all. `ArrayTable` is the *segment table* that holds all the SCI32 arrays. Calling `SegManager::freeArray` deletes one *whole array*.

>In SSCI, this function is named "del"

There is definitely no such “del” function in SSCI for arrays. I can only assume you are very confused and think of a totally different engine.

> This function should only delete a single element from an array, shifting all the rest, and that's what it does

The disassembly for kArrayFree in SSCI is very straightforward, so all I can think is that you didn’t look, or were inadvertently looking at something else:

![screen shot 2017-01-06 at 19 43 51](https://cloud.githubusercontent.com/assets/229244/21738312/7ed382a8-d448-11e6-8cf0-3d551c566d26.png)